### ISSUE TITLE
Local connection error: Failed while connecting to sparklyr to port (8880) for sessionid (103): Gateway in port (8880) did not respond.

### ISSUE THREAD

**Issue by Tazinho** [rstudio/sparklyr#281]
Hi, when I setup sparklyr on windows 10 as described in the RStudio tutorial

```
>devtools::install_github("rstudio/sparkapi", force = TRUE) 
Downloading GitHub repo rstudio/sparkapi@master
from URL https://api.github.com/repos/rstudio/sparkapi/zipball/master
Installing sparkapi
>devtools::install_github("rstudio/sparkapi", force = TRUE) 
Downloading GitHub repo rstudio/sparkapi@master
from URL https://api.github.com/repos/rstudio/sparkapi/zipball/master
Installing sparkapi
"C:/PROGRA~1/R/R-33~1.1/bin/x64/R" --no-site-file --no-environ --no-save --no-restore --quiet CMD INSTALL  \
  "C:/Users/MGO/AppData/Local/Temp/RtmpIz9QMt/devtools1b143eba2958/rstudio-sparkapi-61e3709"  \
  --library="C:/Users/MGO/Documents/R/win-library/3.3" --install-tests 

* installing *source* package 'sparkapi' ...
** R
** inst
** tests
** preparing package for lazy loading
** help
*** installing help indices
** building package indices
** testing if installed package can be loaded
*** arch - i386
*** arch - x64
* DONE (sparkapi)
> library(sparklyr)
> spark_install()
Installing Spark 1.6.2 for Hadoop 2.6 or later.
Downloading from:
- 'https://d3kbcqa49mib13.cloudfront.net/spark-1.6.2-bin-hadoop2.6.tgz'
Installing to:
- 'C:\Users\MGO\AppData\Local\rstudio\spark\Cache/spark-1.6.2-bin-hadoop2.6'
trying URL 'https://d3kbcqa49mib13.cloudfront.net/spark-1.6.2-bin-hadoop2.6.tgz'
Content type 'application/x-tar' length 278057117 bytes (265.2 MB)
downloaded 265.2 MB

Installation complete.
```

I am getting the following error, while locally connecting:

```
> sc <- spark_connect(master = "local")
Created default hadoop bin directory under: C:\Users\MGO\AppData\Local\rstudio\spark\Cache\spark-1.6.2-bin-hadoop2.6\tmp\hadoop
Error in force(code) : 
  Failed while connecting to sparklyr to port (8880) for sessionid (103): Gateway in port (8880) did not respond.
    Path: C:\Users\MGO\AppData\Local\rstudio\spark\Cache\spark-1.6.2-bin-hadoop2.6\bin\spark-submit2.cmd
    Parameters: --class, sparklyr.Backend, --packages, "com.databricks:spark-csv_2.11:1.3.0", "C:\Users\MGO\Documents\R\win-library\3.3\sparklyr\java\sparklyr-1.6-2.10.jar", 8880, 103
    Traceback:
      shell_connection(master = master, spark_home = spark_home, app_name = app_name, version = version, hadoop_version = hadoop_version, shell_args = shell_args, config = config, service = FALSE, extensions = extensions)
      start_shell(master = master, spark_home = spark_home, spark_version = version, app_name = app_name, config = config, jars = spark_config_value(config, "spark.jars.default", list()), packages = spark_config_value(config, "sparklyr.defaultPackages"), extensions = extensions, environment = environment, shell_args = shell_args, service = service)
      tryCatch({
    gatewayInfo <- spark_connect_gateway(gatewayAddress, gatewayPort, sessionId, config = config, isStarting = TRUE)
}, error = function(e) {
    abort_shell(paste("Failed while connecting to sparklyr to port (", gatewayPort, ") for sessionid (", sessionId, "): ", e$message, sep = ""), spark_submit_path, shell_args, output_file, error_file)
})
      tryCatchList(expr, classes, parentenv, handlers)
      tryCatchOne(expr, names, parentenv, handlers[[1]])
      value[[3]](cond)
      abort_shell(paste("Failed while connecting to sparklyr to port (", gatewayPort, ") for sessionid (", sessionId, "): ", e$message, sep = ""), spark_submit_path, shell_args, output_file, error_file)

---- Output Log ----
Ivy Default Cache set to: C:\Users\MGO\.ivy2\cache
The jars for the packages stored in: C:\Users\MGO\.ivy2\jars
:: loading settings :: url = jar:file:/C:/Users/MGO/AppData/Local/rstudio/spark/Cache/spark-1.6.2-bin-hadoop2.6/lib/spark-assembly-1.6.2-hadoop2.6.0.jar!/org/apache/ivy/core/settings/ivysettings.xml
com.databricks#spark-csv_2.11 added as a dependency
:: resolving dependencies :: org.apache.spark#spark-submit-parent;1.0
    confs: [default]
    found com.databricks#spark-csv_2.11;1.3.0 in central
    found org.apache.commons#commons-csv;1.1 in central
    found com.univocity#univocity-parsers;1.5.1 in central
:: resolution report :: resolve 192ms :: artifacts dl 5ms
    :: modules in use:
    com.databricks#spark-csv_2.11;1.3.0 from central in [default]
    com.univocity#univocity-parsers;1.5.1 from central in [default]
    org.apache.commons#commons-csv;1.1 from central in [default]
    ---------------------------------------------------------------------
    |                  |            modules            ||   artifacts   |
    |       conf       | number| search|dwnlded|evicted|| number|dwnlded|
    ---------------------------------------------------------------------
    |      default     |   3   |   0   |   0   |   0   ||   3   |   0   |
    ---------------------------------------------------------------------
:: retrieving :: org.apache.spark#spark-submit-parent
    confs: [default]
    0 artifacts copied, 3 already retrieved (0kB/5ms)

---- Error Log ----
```

My SessionInfo:

```
> sessionInfo()
R version 3.3.1 (2016-06-21)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows >= 8 x64 (build 9200)

locale:
[1] LC_COLLATE=German_Germany.1252  LC_CTYPE=German_Germany.1252    LC_MONETARY=German_Germany.1252
[4] LC_NUMERIC=C                    LC_TIME=German_Germany.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] sparklyr_0.4.19

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.7     digest_0.6.10   dplyr_0.5.0     withr_1.0.2     rprojroot_1.0-2 assertthat_0.1 
 [7] rappdirs_0.3.1  R6_2.2.0        DBI_0.5-1       git2r_0.15.0    magrittr_1.5    httr_1.2.1     
[13] curl_2.1        devtools_1.12.0 config_0.2      tools_3.3.1     yaml_2.1.13     parallel_3.3.1 
[19] memoise_1.0.0   tibble_1.2     
```

Thanks in advance!

**Comment by dsblr** [comment]
I am also getting the similar issues, I have installed the latest version of sparklyr from github.

Error in force(code) : 
  Failed while connecting to sparklyr to port (8880) for sessionid (2044): Gateway in port (8880) did not respond.
    Path: C:\Users\user1\AppData\Local\rstudio\spark\Cache\spark-1.6.2-bin-hadoop2.6\bin\spark-submit2.cmd
    Parameters: --class, sparklyr.Backend, --packages, "com.databricks:spark-csv_2.11:1.3.0", "D:\Users\user1\R\R-3.3.1\library\sparklyr\java\sparklyr-1.6-2.10.jar", 8880, 2044
    Traceback:
      shell_connection(master = master, spark_home = spark_home, app_name = app_name, version = version, hadoop_version = hadoop_version, shell_args = shell_args, config = config, service = FALSE, extensions = extensions)
      start_shell(master = master, spark_home = spark_home, spark_version = version, app_name = app_name, config = config, jars = spark_config_value(config, "spark.jars.default", list()), packages = spark_config_value(config, "sparklyr.defaultPackages"), extensions = extensions, environment = environment, shell_args = shell_args, service = service)
      tryCatch({
    gatewayInfo <- spark_connect_gateway(gatewayAddress, gatewayPort, sessionId, config = config, isStarting = TRUE)
}, error = function(e) {
    abort_shell(paste("Failed while connecting to sparklyr to port (", gatewayPort, ") for sessionid (", sessionId, "): ", e$message, sep = ""), spark_submit_path, shell_args, output_file, error_file)
})
      tryCatchList(expr, classes, parentenv, handlers)
      tryCatchOne(expr, names, parentenv, handlers[[1]])
      value[[3]](cond)
      abort_shell(paste("Failed while connecting to sparklyr to port (", gatewayPort, ") for sessionid (", sessionId, "): ", e$message, sep = ""), spark_submit_path, shell_args, output_file, error_file)

---- Output Log ----
The system cannot find the path specified.

---- Error Log ----

**Comment by dsblr** [comment]
Issue Resolved ---- Output Log ----
 The system cannot find the path specified.

The above error has been solved by changing the JAVA_HOME, Spark don’t recognise the JAVA_HOME, if \bin added to the end of the JAVA_HOME. So I removed the \bin from JAVA_HOME and it started working.

But now I am getting a different error. Please help me on the same.

Error in force(code) : 
 Failed while connecting to sparklyr to port (8880) for sessionid (3787): Gateway in port (8880) did not respond.
 Path: C:\Users\user1\AppData\Local\rstudio\spark\Cache\spark-1.6.2-bin-hadoop2.6\bin\spark-submit2.cmd
 Parameters: --class, sparklyr.Backend, --packages, "com.databricks:spark-csv_2.11:1.3.0", "D:\Users\user1\R\R-3.3.1\library\sparklyr\java\sparklyr-1.6-2.10.jar", 8880, 3787
 Traceback:
 shell_connection(master = master, spark_home = spark_home, app_name = app_name, version = version, hadoop_version = hadoop_version, shell_args = shell_args, config = config, service = FALSE, extensions = extensions)
 start_shell(master = master, spark_home = spark_home, spark_version = version, app_name = app_name, config = config, jars = spark_config_value(config, "spark.jars.default", list()), packages = spark_config_value(config, "sparklyr.defaultPackages"), extensions = extensions, environment = environment, shell_args = shell_args, service = service)
 tryCatch({
 gatewayInfo <- spark_connect_gateway(gatewayAddress, gatewayPort, sessionId, config = config, isStarting = TRUE)
 }, error = function(e) {
 abort_shell(paste("Failed while connecting to sparklyr to port (", gatewayPort, ") for sessionid (", sessionId, "): ", e$message, sep = ""), spark_submit_path, shell_args, output_file, error_file)
 })
 tryCatchList(expr, classes, parentenv, handlers)
 tryCatchOne(expr, names, parentenv, handlers[[1]])
 value[3]
 abort_shell(paste("Failed while connecting to sparklyr to port (", gatewayPort, ") for sessionid (", sessionId, "): ", e$message, sep = ""), spark_submit_path, shell_args, output_file, error_file)

---- Output Log ----
 file:/C:/Users/user1/.m2/repository/com/databricks/spark-csv_2.11/1.3.0/spark-csv_2.11-1.3.0.pom
  -- artifact com.databricks#spark-csv_2.11;1.3.0!spark-csv_2.11.jar:

  file:/C:/Users/user1/.m2/repository/com/databricks/spark-csv_2.11/1.3.0/spark-csv_2.11-1.3.0.jar

==== local-ivy-cache: tried

  C:\Users\user1\.ivy2\local\com.databricks\spark-csv_2.11\1.3.0\ivys\ivy.xml

==== central: tried

  https://repo1.maven.org/maven2/com/databricks/spark-csv_2.11/1.3.0/spark-csv_2.11-1.3.0.pom

  -- artifact com.databricks#spark-csv_2.11;1.3.0!spark-csv_2.11.jar:

  https://repo1.maven.org/maven2/com/databricks/spark-csv_2.11/1.3.0/spark-csv_2.11-1.3.0.jar

==== spark-packages: tried

  http://dl.bintray.com/spark-packages/maven/com/databricks/spark-csv_2.11/1.3.0/spark-csv_2.11-1.3.0.pom

  -- artifact com.databricks#spark-csv_2.11;1.3.0!spark-csv_2.11.jar:

  http://dl.bintray.com/spark-packages/maven/com/databricks/spark-csv_2.11/1.3.0/spark-csv_2.11-1.3.0.jar

    ::::::::::::::::::::::::::::::::::::::::::::::

    ::          UNRESOLVED DEPENDENCIES         ::

    ::::::::::::::::::::::::::::::::::::::::::::::

    :: com.databricks#spark-csv_2.11;1.3.0: not found

    ::::::::::::::::::::::::::::::::::::::::::::::

**Comment by javierluraschi** [comment]
A fix is available in `sparklyr` `0.4.24` that avoids downloading the `com.databricks#spark-csv_2.11;1.3.0` and instead, embeds this resource in the package.

To install the latest version, install `devtools` package and run:

```{r}
devtools::install_github("rstudio/sparklyr")
```

Then retry `spark_connect` and let us know how it goes.

**Comment by Tazinho** [comment]
Thanks a lot!
